src/interpreter/compiler.ts

```diff
@@ -0,0 +1,290 @@
+  type MemberExpression,
+  type Node,
+  type Program,
+  type ReturnStatement,
+  type SequenceExpression,
+  type StringLiteral,
+  type SwitchStatement,
+  type UnaryExpression,
+  type UpdateExpression,
+  type VariableDeclaration,
+  type WhileStatement
+} from '../ast/types';
+import {
+  constructAssignInstr,
+  constructBinaryOperationInstr,
+  constructCallInstr,
+  constructDoneInstr,
+  constructEnterProgramInstr,
+  constructGotoInstr,
+  constructJumpOnFalseInstr,
+  constructLoadConstantInstr,
+  constructLoadFunctionInstr,
+  constructLoadSymbolInstr,
+  constructTeardownInstr,
+  PLACEHOLDER_ADDRESS
+} from './instruction';
+import { isEmptyStatement, isIdentifier } from '../ast/typeGuards';
+import { isNotUndefined } from '../utils/typeGuards';
+import {
+  InvalidCallError,
+  UnsupportedOperatorError,
+  UnsupportedOperatorErrorType
+} from './errors';
+import {
+  constructConditionalExpression,
+  constructFalseConstant,
+  constructMainCallExpression,
+  constructTrueConstant
+} from '../ast/constructors';
+import { type SymbolTable } from './types/symbolTable';
+import {
+  addBlockSymbolTableEntries,
+  addFunctionSymbolTableEntries,
+  addProgramSymbolTableEntries,
+  getFunctionSymbolTableEntry,
+  getNumOfEntriesInFrame,
+  getSymbolTableEntry,
+  getSymbolTableEntryInFrame
+} from './symbolTable';
+import { type Instr } from './types/instruction';
+
+export const compileProgram = (ast: Program): Instr[] => {
+  const symbolTable: SymbolTable = {
+    head: {},
+    tail: null,
+    parent: null
+  };
+  const state: CompilerState = {
+    symbolTable,
+    instructions: []
+  };
+  compile(ast, state);
+  const mainCallExpression = constructMainCallExpression();
+  compile(mainCallExpression, state);
+  const doneInstr = constructDoneInstr();
+  state.instructions.push(doneInstr);
+  return state.instructions;
+};
+
+const compile = (node: Node, state: CompilerState): void => {
+  // The typecast allows for mapping to a specific evaluator instr type from their union type.
+  // https://stackoverflow.com/questions/64527150/in-typescript-how-to-select-a-type-from-a-union-using-a-literal-type-property
+  compilers[node.type](node as any, state);
+};
+
+const compilers: CompilerMapping = {
+  ArrayAccessExpression: (
+    node: ArrayAccessExpression,
+    state: CompilerState
+  ) => {},
+  AssignmentExpression: (
+    node: AssignmentExpression,
+    state: CompilerState
+  ) => {},
+  BinaryExpression: (node: BinaryExpression, state: CompilerState) => {
+    compile(node.left, state);
+    compile(node.right, state);
+    const binaryOperationInstr = constructBinaryOperationInstr(node.operator);
+    state.instructions.push(binaryOperationInstr);
+  },
+  BlockStatement: (node: BlockStatement, state: CompilerState) => {
+    const blockSymbolTable = addBlockSymbolTableEntries(
+      node,
+      state.symbolTable
+    );
+    node.items.forEach((item) => {
+      state.symbolTable = blockSymbolTable;
```

Comment:
Can this statement be shifted out of the loop? Or is it possible that other compilers will change the value of `state.symbolTable`?

If the answer to the second question above is yes, does the value of `state.symbolTable` matter when exiting this compiler function? Should it be set to anything in particular?